stan/math/gpu/kernel_cl.hpp

```diff
@@ -0,0 +1,267 @@
+  const char* transpose_matrix =
+#include <stan/math/gpu/kernels/transpose_matrix.cl>  // NOLINT
+      ;                                               // NOLINT
+  const char* zeros_matrix =
+#include <stan/math/gpu/kernels/zeros_matrix.cl>  // NOLINT
+      ;                                           // NOLINT
+  const char* identity_matrix =
+#include <stan/math/gpu/kernels/identity_matrix.cl>  // NOLINT
+      ;                                              // NOLINT
+  const char* copy_triangular_matrix =
+#include <stan/math/gpu/kernels/copy_triangular_matrix.cl>  // NOLINT
+      ;                                                     // NOLINT
+  const char* copy_triangular_transposed_matrix =
+#include <stan/math/gpu/kernels/triangular_transpose.cl>  // NOLINT
+      ;                                                   // NOLINT
+  const char* copy_submatrix =
+#include <stan/math/gpu/kernels/sub_block.cl>  // NOLINT
+      ;                                        // NOLINT
+  const char* check_nan =
+#include <stan/math/gpu/kernels/check_nan.cl>  // NOLINT
+      ;                                        // NOLINT
+  const char* check_diagonal_zeros =
+#include <stan/math/gpu/kernels/check_diagonal_zeros.cl>  // NOLINT
+      ;                                                   // NOLINT
+  const char* check_symmetric =
+#include <stan/math/gpu/kernels/check_symmetric.cl>  // NOLINT
+      ;                                              // NOLINT
+  const char* subtract_symmetric =
+#include <stan/math/gpu/kernels/subtract_matrix.cl>  // NOLINT
+      ;                                              // NOLINT
+  const char* add_symmetric =
+#include <stan/math/gpu/kernels/add_matrix.cl>  // NOLINT
+      ;                                         // NOLINT
+
+ protected:
+  const char* helpers =
+#include <stan/math/gpu/kernels/helpers.cl>  // NOLINT
+      ;                                      // NOLINT
+  // Holds Default parameter values for each Kernel.
+  typedef std::map<const char*, int> map_base_opts;
+  const map_base_opts base_opts = {{"LOWER", gpu::Lower},
+                                   {"UPPER", gpu::Upper},
+                                   {"ENTIRE", gpu::Entire},
+                                   {"UPPER_TO_LOWER", gpu::UpperToLower},
+                                   {"LOWER_TO_UPPER", gpu::LowerToUpper}};
+
+  /** Holds meta information about a kernel.
+   * @param exists a bool to identify whether a kernel has been compiled.
+   * @param group The name of the compilation group for the kernel.
+   * @param opts The compile time options for the kernel.
+   * @param code The source code for the kernel.
+   */
+  struct kernel_meta_info {
+    bool exists;
+    std::vector<const char*> opts;
+    const char* raw_code;
+  };
+  /**
+   * Map of a kernel name (first) and it's meta information (second).
+   */
+  typedef std::map<const char*, kernel_meta_info> map_table;
+  const map_table kernel_table
```

Comment:
Redundant because the `kernel_cl` appears twice on that line? I think the kernels should live on some namespace and it might as well also be the kernel_cl one. The benefit to a call like that is that we get rid of all of these `std::map` and stringly typed code and have the compiler check that we're trying to instantiate an actual kernel. Also I've heard the `std::map`s are surprisingly hard to get right+performant in a threaded context, though I suppose most of this GPU stuff won't work with threads enabled yet. But I think the code will be a little more simple and statically typed, at the expense of a user typing `kernel_cl` one extra time. But ironically the entire thing can be autocompleted with the static version because the compiler can tell what your available kernels are :)

Reply:
I'm buying this idea. I just changed the `kernel_cl_base` class so that all the kernels are compiled on first construction so tomorrow I'll play with this. Before you were saying we would have a struct with the kernel code and meta info. But now the map is just `'name' : kernel_code`. Maybe we can do something like

```cpp
kernel_cl.identity(A.buffer(), A.rows(), A.cols());
```

Reply:
These could also set in an opencl namespace so it would be 
```
stan::math::opencl::identity(A.buffer(), A.rows(), A.cols())
```

Reply:
Great! This looks awesome:
> `kernel_cl.identity(A.buffer(), A.rows(), A.cols());`

I like that this looks much more like a normal C++ method. I think this signature is made less redundant now (now that we are precompiling) because you don't have to put the dynamic compile check at the beginning of each of the functions, though I suppose the setArgs stuff might be semi-duplicated inside this version. But better duplicated inside the API than having users do it, I think.

I kind of prefer a class in `stan::math` rather than creating a new namespace slightly stylistically but also it's organizationally nice if these functions live on the class whose internal data they need to access to operate. It may also allow more performant code by allowing less calls to `getInstance()`.

Reply:
Cool I didn't have time to finish this before work but I'm setting it up like 

```cpp
  auto identity(matrix_gpu A)
      : rows_(A.rows()), cols_(A.cols()), buffer_(A.buffer()) {
    cl::Kernel kern = this->kernels()[("identity")];
    this->set_args(kern, buffer_, rows_, cols_);
    return kern;
  }
```
(where `kernels()` is)
```cpp
  inline auto kernels() {
    return internal::kernel_cl_base::getInstance().kernels;
  }
```

Is there a way to get rid of that map of kernels from the `kernel_cl_base`?

Reply:
So I couldn't make it as nice as the above, but I added methods to `kernel_cl` so users can call 

```cpp
auto sub_kernel = kernel_cl.subtract(C.buffer(), A.buffer(), B.buffer(), A.rows(), A.cols());
```